From: Stefan Eissing <stefan@eissing.org>
Date: Mon, 11 May 2026 14:56:04 +0200
Subject: event: fix wakeup consumption

The events on a multi wakeup socketpair were only consumed via
curl_multi_poll()/curl_multi_wait() but not in event based processing on
a curl_multi_socket() call. That led to busy loops as reported in

Fixes #21547
Reported-by: Earnestly on github
Closes #21549

Origin: upstream, https://github.com/curl/curl/commit/2a2104f3cff44bb28bb570a093be52bbeeed8f23
Bug: https://github.com/curl/curl/issues/21547
Bug-Debian: https://bugs.debian.org/1136378
Last-Update: 2025-05-13
---
 lib/multi.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/multi.c b/lib/multi.c
index 7520253..a0012d8 100644
--- a/lib/multi.c
+++ b/lib/multi.c
@@ -2479,6 +2479,11 @@ static CURLMcode multi_runsingle(struct Curl_multi *multi,
   Curl_uint32_bset_remove(&multi->dirty, data->mid);
 
   if(data == multi->admin) {
+#ifdef ENABLE_WAKEUP
+    /* Consume any pending wakeup signals before processing.
+     * This is necessary for event based processing. See #21547 */
+    (void)Curl_wakeup_consume(multi->wakeup_pair, TRUE);
+#endif
 #ifdef USE_RESOLV_THREADED
     Curl_async_thrdd_multi_process(multi);
 #endif
